stages/pod-general.yaml

```diff
@@ -0,0 +1,357 @@
+kind: Stage
+apiVersion: kwok.x-k8s.io/v1alpha1
+metadata:
+  name: pod-create
+spec:
+  resourceRef:
+    apiGroup: v1
+    kind: Pod
+  selector:
+    matchExpressions:
+      - key: '.metadata.deletionTimestamp'
+        operator: 'DoesNotExist'
+      - key: '.status.podIP'
+        operator: 'DoesNotExist'
+  weight: 1
+  delay:
+    durationSeconds: 1
+  next:
+    event:
+      type: Normal
+      reason: Created
+      message: Created container
+    finalizers:
+      add:
+        - 'kwok.x-k8s.io/fake'
+    statusTemplate: |
+      {{ $now := Now }}
+      conditions:
+      {{ if .spec.initContainers }}
+        - lastProbeTime: null
+          lastTransitionTime: '{{ $now }}'
+          message: 'containers with incomplete status: [{{ range .spec.initContainers }} {{ .name }} {{ end }}]'
+          reason: ContainersNotInitialized
+          status: "False"
+          type: Initialized
+      {{ else }}
+        - lastProbeTime: null
+          lastTransitionTime: '{{ $now }}'
+          status: "True"
+          type: Initialized
+      {{ end }}
+        - lastProbeTime: null
+          lastTransitionTime: '{{ $now }}'
+          message: 'containers with unready status: [{{ range .spec.containers }} {{ .name }} {{ end }}]'
+          reason: ContainersNotReady
+          status: "False"
+          type: Ready
+        - lastProbeTime: null
+          lastTransitionTime: '{{ $now }}'
+          message: 'containers with unready status: [{{ range .spec.containers }} {{ .name }} {{ end }}]'
+          reason: ContainersNotReady
+          status: "False"
+          type: ContainersReady
+      {{ range .spec.readinessGates }}
+        - lastTransitionTime: {{ $now }}
+          status: "True"
+          type: {{ .conditionType }}
+      {{ end }}
+
+      {{ if .spec.initContainers }}
+      initContainerStatuses:
+        {{ range .spec.initContainers }}
+        - image: {{ .image }}
+          name: {{ .name }}
+          ready: false
+          restartCount: 0
+          started: false
+          state:
+            waiting:
+              reason: PodInitializing
+        {{ end }}
+      containerStatuses:
+        {{ range .spec.containers }}
+        - image: {{ .image }}
+          name: {{ .name }}
+          ready: false
+          restartCount: 0
+          started: false
+          state:
+            waiting:
+              reason: PodInitializing
+        {{ end }}
+      {{ else }}
+      containerStatuses:
+        {{ range .spec.containers }}
+        - image: {{ .image }}
+          name: {{ .name }}
+          ready: false
+          restartCount: 0
+          started: false
+          state:
+            waiting:
+              reason: ContainerCreating
+        {{ end }}
+      {{ end }}
+      hostIP: {{ with .status.hostIP }} {{ . }} {{ else }} {{ NodeIP }} {{ end }}
+      podIP: {{ with .status.podIP }} {{ . }} {{ else }} {{ PodIP }} {{ end }}
+      phase: Pending
+---
+kind: Stage
```

Comment:
From my point of view, it would be a little too complex for end-users to config based on current implementations, providing examples mitigate this a little but not much I think, especially for someone who're SREs not developers.

Besides, I would like to see something more like this, but I have no idea whether this meets all your needs and descriptive enough. This can also not a CRD I think. 
``` yaml
Steps:
- name: create pod
   phase: Pending
   filePath: <path-to-pod.yaml>
   condition:
     type:
     status:
     lastProbeTime:
     lastTransitionTime:
     reason:
     message:
- name: running pod
   after: 10s # waiting time after prev step
   phase: Running
   condition:
- name: liveness probe error # something like this
   phase: Fail
   condition:
- name: delete pod
   phase: Terminating
   condition:
- name: pod is terminated
   phase: Terminated
# - phase: Succeed
# - phase: Unknown
```

This is declarative and if necessary, you can export pod field if you like.

Reply:
https://github.com/kubernetes-sigs/kwok/issues/67
https://github.com/kubernetes-sigs/kwok/issues/67#issuecomment-1312732001

This is my first version of the design, which has the fatal problem of needing to have a state to maintain the stage where the Pod is, 
and if the kwok is restarted, it can't pick up where it left off.